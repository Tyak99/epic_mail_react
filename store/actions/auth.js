import axios from 'axios';
import * as actionTypes from '../actionConstants';

const url = 'https://intense-thicket-60071.herokuapp.com/api/v1/auth';
export const authStart = () => ({
  type: actionTypes.AUTH_START,
});

export const authSuccess = data => ({
  type: actionTypes.AUTH_SUCCESS,
  data,
});

export const authFailed = error => ({
  type: actionTypes.AUTH_FAILED,
  error,
});
export const login = (email, password) => {
  const authData = {
    email,
    password,
  };
  return (dispatch) => {
    dispatch(authStart());
    return axios
      .post(`${url}/login`, authData)
      .then((response) => {
        const { token } = response.data.data;
        localStorage.removeItem('token');
        localStorage.setItem('token', token);
        dispatch(authSuccess(response.data.data));
      })
      .catch((error) => {
        dispatch(authFailed(error.response.data.error));
      });
  };
};

export const register = authData => (dispatch) => {
  dispatch(authStart());
  const regex = /^[a-z0-9][a-z0-9-_]+@([a-z]|[a-z0-9]?[a-z0-9-]+[a-z0-9])\.[a-z0-9]{2,10}(?:\.[a-z]{2,10})?$/;
  if (authData.password !== authData.confirmPassword) {
    return dispatch(authFailed('Error! Passwords do not match'));
  }
  if (!regex.test(authData.email)) {
    return dispatch(authFailed('Error! Invalid email address provided'));
  }
  return axios.post(`${url}/signup`, authData)
    .then((res) => {
      const { token } = res.data.data;
      localStorage.removeItem('token');
      localStorage.setItem('token', token);
      dispatch(authSuccess(res.data.data));
    })
    .catch((error) => {
      dispatch(authFailed(error.response.data.error));
    });
};

export const checkAuth = () => (dispatch) => {
  const token = localStorage.getItem('token');
  const name = localStorage.getItem('name');
  if (token) {
    dispatch(authSuccess(token, name));
  }
};

export const logoutUser = () => {
  localStorage.removeItem('token');
  return {
    type: 'LOGOUT',
  };
};
