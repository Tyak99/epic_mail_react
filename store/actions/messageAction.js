import axios from 'axios';
import * as actionTypes from '../actionConstants';

const url = 'https://intense-thicket-60071.herokuapp.com/api/v1/messages';
export const getMessageSuccess = data => ({
  type: actionTypes.GET_INBOX_SUCCESS,
  data,
});

export const getSentMessageSuccess = data => ({
  type: actionTypes.GET_SENT_SUCCESS,
  data,
});

export const getMessageFailed = error => ({
  type: actionTypes.GET_MESSAGE_FAILED,
  error,
});

export const sendMessageFailed = error => ({
  type: actionTypes.SEND_MESSAGE_FAILED,
  error,
});

export const sendMessageStart = () => ({
  type: actionTypes.SEND_MESSAGE_START,
});

export const getSingleMessageSuccess = message => ({
  type: 'GET_SINGLE_MESSAGE',
  message,
});

export const sendMessageSuccess = () => ({
  type: actionTypes.SEND_MESSAGE_SUCCESS,
  message: 'Sent successfully',
});

export const getInbox = () => (dispatch) => {
  axios
    .get(url, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    })
    .then((res) => {
      dispatch(getMessageSuccess(res.data.data));
    })
    .catch((error) => {
      dispatch(getMessageFailed(error.response.data.error));
    });
};

export const getSentMessages = () => (dispatch) => {
  axios
    .get(`${url}/sent`, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    })
    .then((res) => {
      dispatch(getSentMessageSuccess(res.data.data));
    })
    .catch((error) => {
      dispatch(getMessageFailed(error.response.data.error));
    });
};

const validateEmail = (data) => {
  const emailCheck = /\S+@\S+\.\S+/;
  return emailCheck.test(data);
};

export const sendMessage = (data) => {
  const checkEmail = validateEmail(data.emailTo);
  const config = {
    headers: {
      Authorization: `${localStorage.getItem('token')}`,
    },
  };
  return (dispatch) => {
    if (checkEmail === true) {
      dispatch(sendMessageStart());
      return axios.post(url, data, config)
        .then((res) => {
          dispatch(sendMessageSuccess(res));
        })
        .catch((error) => {
          dispatch(sendMessageFailed(error.response.data.error));
        });
    }
    return dispatch(sendMessageFailed('Invalid email'));
  };
};

export const getSingleMessage = id => (dispatch) => {
  axios.get(`${url}/${id}`, {
    headers: { Authorization: `${localStorage.getItem('token')}` },
  })
    .then((res) => {
      dispatch(getSingleMessageSuccess(res.data.data));
    })
    .catch((error) => {
      dispatch(getMessageFailed(error.response.data.error));
    });
};
