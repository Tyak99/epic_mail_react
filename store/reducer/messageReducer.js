import * as actionTypes from '../actionConstants';

export const initialState = {
  inbox: [],
  sent: [],
  error: '',
  isLoading: false,
  status: '',
};
export default (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.GET_INBOX_SUCCESS:
      return {
        ...state,
        inbox: action.data,
      };
    case actionTypes.GET_MESSAGE_FAILED:
      return {
        ...state,
        error: action.error,
      };
    case actionTypes.GET_SENT_SUCCESS:
      return {
        ...state,
        sent: action.data,
      };
    case actionTypes.SEND_MESSAGE_START:
      return {
        ...state,
        isLoading: true,
      };
    case actionTypes.SEND_MESSAGE_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: '',
        status: action.message,
      };
    case actionTypes.SEND_MESSAGE_FAILED:
      return {
        ...state,
        isLoading: false,
        status: '',
        error: action.error,
      };
    default:
      return state;
  }
};
