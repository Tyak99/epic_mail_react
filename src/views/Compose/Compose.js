import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Navigation from '../../components/Navigation/Navigation';
import ComposeForm from '../../components/ComposeForm/ComposeForm';
import * as actionCreators from '../../../store/actions/messageAction';

export class Compose extends Component {
  state = {
    to: '',
    subject: '',
    body: '',
  }

  handleInput = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  submitHandler = () => {
    const { to, subject, body } = this.state;
    const { sendMessage } = this.props;
    const data = {
      emailTo: to,
      subject,
      message: body,
    };
    sendMessage(data);
  }

  render() {
    const { isLoading, error, status } = this.props;
    return (
      <Fragment>
        <Navigation />
        <ComposeForm
          handleInput={this.handleInput}
          submitHandler={this.submitHandler}
          loading={isLoading}
          error={error}
          notification={status}
        />
      </Fragment>
    );
  }
}

Compose.defaultProps = {
  isLoading: false,
  error: null,
  status: '',
};

Compose.propTypes = {
  sendMessage: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  status: PropTypes.string,
};

const mapStateToProps = state => ({
  isLoading: state.messages.isLoading,
  error: state.messages.error,
  status: state.messages.status,
});
const mapDispatchToProps = dispatch => ({
  sendMessage: data => dispatch(actionCreators.sendMessage(data)),
});
export default connect(mapStateToProps, mapDispatchToProps)(Compose);
