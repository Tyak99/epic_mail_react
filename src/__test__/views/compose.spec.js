import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import configureMockStore from 'redux-mock-store';
import nock from 'nock';
import thunk from 'redux-thunk';
import { Compose } from '../../views/Compose/Compose';
import { sendMessageSuccess, sendMessageFailed, sendMessage } from '../../../store/actions/messageAction';
import messageReducer, { initialState } from '../../../store/reducer/messageReducer';


const props = {
  error: '',
  isLoading: false,
  sendMessage: () => {},
};

describe('component: Compose', () => {
  it('should render correctly', (done) => {
    shallow(<Compose {...props} />);
    done();
  });
//   it('should render the compose form component', (done) => {
//     const wrapper = shallow(<Compose {...props} />);
//     expect(wrapper.find('ComposeForm').shallow()).toBe(true);
//     done();
//   });
});

describe('Compose Form', () => {
  let wrapper;
  let inputHandler;
  let submitHandler;
  const mockEvent = { target: { name: 'email', value: 'email' } };
  beforeEach(() => {
    wrapper = shallow(<Compose {...props} />);
    inputHandler = sinon.spy(wrapper.instance(), 'handleInput');
    submitHandler = sinon.spy(wrapper.instance(), 'submitHandler');
    wrapper.instance().forceUpdate();
  });
  it('should call the handleInput method when the input is updated', (done) => {
    const emailInput = wrapper.find('ComposeForm').shallow().find('.email-to');
    expect(emailInput.length).toBe(1);
    emailInput.simulate('change', mockEvent);
    expect(inputHandler.called).toBe(true);
    done();
  });
  it('should call the submit handler when the send button is clicked', (done) => {
    const sendButton = wrapper.find('ComposeForm').shallow().find('#btn-send');
    expect(sendButton.length).toBe(1);
    sendButton.simulate('click');
    expect(submitHandler.called).toBe(true);
    done();
  });
});

describe('send message actions', () => {
  const mockStore = configureMockStore();
  const store = mockStore({});
  it('dispatches send message success when message sent', (done) => {
    store.dispatch(sendMessageSuccess());
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
  it('dispatches send message faled when message fails', (done) => {
    store.dispatch(sendMessageFailed('Invalid email'));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
});

describe('send message async actions', () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const mockData = {
    emailTo: 'tunde@mail.com',
    subject: 'Hello',
    message: 'Please save me',
  };
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });
  it('sends message successfully', (done) => {
    nock('https://intense-thicket-60071.herokuapp.com/api/v1')
      .post('/messages')
      .reply(201, {});
    return store.dispatch(sendMessage(mockData)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
  it('calls send message failed', (done) => {
    nock('https://intense-thicket-60071.herokuapp.com/api/v1')
      .post('/messages')
      .reply(422, {});
    return store.dispatch(sendMessage({ emailTo: 'tunde@mail.com' })).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
});

describe('message reducer', () => {
  it('should return the initial state', (done) => {
    expect(messageReducer(undefined, {})).toMatchSnapshot();
    done();
  });
  it('should update the state on sendMessage start', (done) => {
    expect(messageReducer(initialState, { type: 'SEND_MESSAGE_START' })).toMatchSnapshot();
    done();
  });
  it('should update the state on sendMessage success', (done) => {
    expect(messageReducer(initialState, { type: 'SEND_MESSAGE_SUCCESS' })).toMatchSnapshot();
    done();
  });
  it('should update the state on sendMessage failed', (done) => {
    expect(messageReducer(initialState, { type: 'SEND_MESSAGE_FAILED' })).toMatchSnapshot();
    done();
  });
});
