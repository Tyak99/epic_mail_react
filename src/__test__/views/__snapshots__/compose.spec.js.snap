// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message reducer should return the initial state 1`] = `
Object {
  "error": "",
  "inbox": Array [],
  "isLoading": false,
  "sent": Array [],
  "status": "",
}
`;

exports[`message reducer should update the state on sendMessage failed 1`] = `
Object {
  "error": undefined,
  "inbox": Array [],
  "isLoading": false,
  "sent": Array [],
  "status": "",
}
`;

exports[`message reducer should update the state on sendMessage start 1`] = `
Object {
  "error": "",
  "inbox": Array [],
  "isLoading": true,
  "sent": Array [],
  "status": "",
}
`;

exports[`message reducer should update the state on sendMessage success 1`] = `
Object {
  "error": "",
  "inbox": Array [],
  "isLoading": false,
  "sent": Array [],
  "status": undefined,
}
`;

exports[`send message actions dispatches send message faled when message fails 1`] = `
Array [
  Object {
    "message": "Sent successfully",
    "type": "SEND_MESSAGE_SUCCESS",
  },
  Object {
    "error": "Invalid email",
    "type": "SEND_MESSAGE_FAILED",
  },
]
`;

exports[`send message actions dispatches send message success when message sent 1`] = `
Array [
  Object {
    "message": "Sent successfully",
    "type": "SEND_MESSAGE_SUCCESS",
  },
]
`;

exports[`send message async actions calls send message failed 1`] = `
Array [
  Object {
    "type": "SEND_MESSAGE_START",
  },
  Object {
    "error": undefined,
    "type": "SEND_MESSAGE_FAILED",
  },
]
`;

exports[`send message async actions sends message successfully 1`] = `
Array [
  Object {
    "type": "SEND_MESSAGE_START",
  },
  Object {
    "message": "Sent successfully",
    "type": "SEND_MESSAGE_SUCCESS",
  },
]
`;
